staging/src/k8s.io/client-go/examples/workqueue/main.go

```diff
@@ -0,0 +1,187 @@
+		glog.Fatal("Timed out waiting for caches to sync")
+	}
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Info("Stopping Pod controller")
+}
+
+func (c *Controller) runWorker() {
+	for c.f(c.indexer, c.queue) {
+	}
+}
+
+func main() {
+	var kubeconfig string
+	var master string
+
+	flag.StringVar(&kubeconfig, "kubeconfig", "", "absolute path to the kubeconfig file")
+	flag.StringVar(&master, "master", "", "master url")
+	flag.Parse()
+
+	// Create the connection
+	config, err := clientcmd.BuildConfigFromFlags(master, kubeconfig)
+	if err != nil {
+		glog.Fatal(err)
+	}
+
+	// creates the clientset
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		glog.Fatal(err)
+	}
+
+	// create the pod watcher
+	podListWatcher := cache.NewListWatchFromClient(clientset.Core().RESTClient(), "pods", api.NamespaceDefault, fields.Everything())
+
+	// create the workqueue
+	queue := workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter())
+
+	// Bind the workqueue to a cache with the help of an informer. This way we make sure than whenever the cache
+	// is updated, the pod key is added to the workqueue. Note than when we finally process the item from the
+	// workqueue we might see a newer version of the Pod than the version which was responsible for triggering the update.
+	indexer, informer := cache.NewIndexerInformer(podListWatcher, &v1.Pod{}, 0, cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(obj)
+			if err == nil {
+				queue.Add(key)
+			}
+		},
+		UpdateFunc: func(old interface{}, new interface{}) {
+			key, err := cache.MetaNamespaceKeyFunc(new)
+			if err == nil {
+				queue.Add(key)
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			// IndexerInformer uses a delta queue, therefore for deletes we have to use this
+			// key function.
+			key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+			if err == nil {
+				queue.Add(key)
+			}
+		},
+	}, cache.Indexers{})
+
+	controller := NewController(queue, indexer, informer, func(indexer cache.Indexer, queue workqueue.RateLimitingInterface) bool {
+		// Wait until there is a new item in the working queue
+		key, quit := queue.Get()
+		if quit {
+			return false
+		}
+		// Tell the queue that we are done with processing this key. This unblocks the key for other workers
+		// This allows safe parallel processing because two pods with the same key are never processed in
+		// parallel.
+		defer queue.Done(key)
+		// Fetch the latest Pod state from cache
+		obj, exists, err := indexer.GetByKey(key.(string))
+
+		if err != nil {
+			// TODO: Does this make sense?
+			// Tricky what to do in this situation. One thing we can do, is enqueueing it a few times to
+			// add some backoff delays on the invalid key. This way we avoid hotlooping
+			// on invalid keys.
+			if queue.NumRequeues(key) < 5 {
```

Comment:
I can't see where the business logic of your controller exists. As a convention we use a separate function for it which surfaces an error at this level and then another function that is your error handling code which also facilitates requeues vs drops of objects. See #8 in https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md